File: leetCode/tree/medium/quiz102/java/SolutionV1.java
package tree.medium.quiz102.java;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class SolutionV1 {

    public static void main(String[] args) {
        TreeNode n9 = new TreeNode(9);
        TreeNode n15 = new TreeNode(15);
        TreeNode n7 = new TreeNode(7);
        TreeNode n20 = new TreeNode(20, n15, n7);
        TreeNode root = new TreeNode(3, n9, n20);

        List<List<Integer>> answer = new SolutionV1().levelOrder(root);
        for (List<Integer> integers : answer) {
            System.out.println(integers.toString());
        }
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;

        Queue<TreeNode> que = new LinkedList<>();
        que.add(root);

        while (!que.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            int cnt = que.size();

            for (int i = 0; i < cnt; i++) {
                TreeNode node = que.poll();
                level.add(node.val);
                if (node.left != null) {
                    que.add(node.left);
                }
                if (node.right != null) {
                    que.add(node.right);
                }
            }
            res.add(level);
        }
        return res;
    }
}

File: leetCode/tree/medium/quiz102/kotlin/SolutionV1.kt
package tree.medium.quiz102.kotlin

import tree.medium.quiz102.kotlin.SolutionV1.TreeNode

fun main() {
    val n9 = TreeNode(9)
    val n15 = TreeNode(15)
    val n7 = TreeNode(7)
    val n20 = TreeNode(20).apply {
        left = n15
        right = n7
    }
    val root = TreeNode(3).apply {
        left = n9
        right = n20
    }

    SolutionV1().levelOrder(root).forEach { println(it.toString()) }
}

class SolutionV1 {
    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }

    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val res = mutableListOf<MutableList<Int>>()
        if (root == null) return res

        val que = ArrayDeque<TreeNode>()
        que.addLast(root)

        while (que.isNotEmpty()) {
            val level = mutableListOf<Int>()

            for (i in 0 until que.size) {
                val node = que.removeFirst()
                level.add(node.`val`)

                if (node.left != null) {
                    que.addLast(node.left!!)
                }

                if (node.right != null) {
                    que.addLast(node.right!!)
                }
            }
            res.add(level)
        }
        return res
    }
}
